Подход: можно использовать churn rate, чтобы оценить отток пользователей в 2016 и 2017 году. Также может быть полезен retentation rate, чтобы узнать, как много клиентов всё ещё остаётся в 2017 году и сравнить с показателями 2016 года

In [44]:
import pandas as pd
pd.__version__

'1.4.0'

In [53]:
orders = pd.read_csv(
    'Studies/orders.csv'
)
orders

id  order_date ship_mode customer_id     sales
0     100006  2014-09-07  Standard    DK-13375   377.970
1     100090  2014-07-08  Standard    EB-13705   699.192
2     100293  2014-03-14  Standard    NF-18475    91.056
3     100328  2014-01-28  Standard    JC-15340     3.928
4     100363  2014-04-08  Standard    JM-15655    21.376
...      ...         ...       ...         ...       ...
5004  168802  2017-11-03  Standard    JO-15145    18.368
5005  169320  2017-07-23    Second    LH-16900   171.430
5006  169488  2017-09-07     First    AA-10375    56.860
5007  169502  2017-08-28  Standard    MG-17650   113.410
5008  169551  2017-07-07     First    RL-19615  1344.838

[5009 rows x 5 columns]

In [27]:
customers = pd.read_csv(
    'Studies/customers.csv'
)
customers

id               name      segment           state  \
0    CG-12520        Claire Gute     Consumer        Kentucky   
1    DV-13045    Darrin Van Huff    Corporate      California   
2    SO-20335     Sean O'Donnell     Consumer         Florida   
3    BH-11710    Brosina Hoffman     Consumer      California   
4    AA-10480       Andrew Allen     Consumer  North Carolina   
..        ...                ...          ...             ...   
788  CJ-11875       Carl Jackson    Corporate    Pennsylvania   
789  RS-19870         Roy Skaria  Home Office            Iowa   
790  SC-20845         Sung Chung     Consumer        Illinois   
791  RE-19405    Ricardo Emerson     Consumer            Ohio   
792  SM-20905  Susan MacKendrick     Consumer            Ohio   

                  city  
0            Henderson  
1          Los Angeles  
2      Fort Lauderdale  
3          Los Angeles  
4              Concord  
..                 ...  
788       Philadelphia  
789         Burlington  
790  Arlington Heights  
791               Kent  
792             Newark  

[793 rows x 5 columns]

Предобработка данных: убираем дубликаты и пустые значения 

In [54]:
orders.drop_duplicates(inplace=True)
customers.drop_duplicates(inplace=True)

In [55]:
orders=orders[(orders.customer_id.notnull()) & (orders.sales!=0)]

In [56]:
orders['order_date'] = pd.to_datetime(orders['order_date'], infer_datetime_format=True)
orders['order_date'] = orders['order_date'].dt.strftime('%Y-%m') 

Разедлим данные на 2016 год и 2017 и проанализируем кол-во пользователей 

In [57]:
orders2016 = orders.loc[(orders['order_date'] >= '2016-01')
                     & (orders['order_date'] < '2017-01')]
orders2016

id order_date ship_mode customer_id     sales
1699  100041    2016-11  Standard    BF-10975   328.540
1700  100083    2016-11  Standard    CD-11980    24.784
1701  100153    2016-12  Standard    KH-16630    63.880
1702  100244    2016-09  Standard    GM-14695   475.694
1703  100300    2016-06    Second    MJ-17740  4823.090
...      ...        ...       ...         ...       ...
4715  168620    2016-12    Second    RB-19795   823.610
4716  169040    2016-12  Standard    GT-14710  1169.454
4717  169369    2016-12  Standard    EC-14050   299.070
4745  105830    2016-12     First    DB-13660   156.512
4990  165953    2016-12  Standard    EP-13915    71.096

[1315 rows x 5 columns]

In [58]:
orders2017 = orders.loc[(orders['order_date'] >= '2017-01')
                     & (orders['order_date'] <= '2017-12')]
orders2017

id order_date ship_mode customer_id     sales
2794  100013    2017-11  Standard    ZC-21910    29.690
2795  100055    2017-05  Standard    MD-17860   152.590
2796  100097    2017-11    Second    MN-17935  1115.250
2797  100111    2017-09  Standard    SV-20365  7359.918
2798  100160    2017-09  Standard    CB-12025    49.688
...      ...        ...       ...         ...       ...
5004  168802    2017-11  Standard    JO-15145    18.368
5005  169320    2017-07    Second    LH-16900   171.430
5006  169488    2017-09     First    AA-10375    56.860
5007  169502    2017-08  Standard    MG-17650   113.410
5008  169551    2017-07     First    RL-19615  1344.838

[1687 rows x 5 columns]

Здесь я смотрю, насколько в целом заказов в 2017 году меньше, чем в 2016 году

In [73]:
count2016 = orders.loc[(orders['order_date'] >= '2016-01')
                     & (orders['order_date'] <= '2016-12')]['id'].count()
count2017 = orders.loc[(orders['order_date'] >= '2017-01')
                     & (orders['order_date'] <= '2017-12')]['id'].count()
churn_rate_orders = count2017 - count2016
churn_rate_orders

372

Создадим таблицы с количеством клиентов по месяцам в 2016 и в 2017 году

In [116]:
orders2016_cnt = orders2016.groupby(['order_date'], as_index=False )['customer_id'].nunique()
orders2016_cnt

order_date  customer_id
0     2016-01           46
1     2016-02           42
2     2016-03           80
3     2016-04           83
4     2016-05           96
5     2016-06           90
6     2016-07           89
7     2016-08           86
8     2016-09          176
9     2016-10           95
10    2016-11          165
11    2016-12          157

In [114]:
orders2017_cnt = orders2017.groupby(['order_date'], as_index=False )['customer_id'].nunique()
orders2017_cnt

order_date  customer_id
0     2017-01           67
1     2017-02           53
2     2017-03          115
3     2017-04          109
4     2017-05          113
5     2017-06          118
6     2017-07          102
7     2017-08          104
8     2017-09          196
9     2017-10          136
10    2017-11          216
11    2017-12          195

In [117]:
prev_cnt = 157
for index, row in orders2017_cnt.iterrows():
    cnt = (prev_cnt - row['customer_id'])/prev_cnt
    print(f'Churn rate at month {index + 1}: {cnt:.2f}')
    prev_cnt = row['customer_id']

Churn rate at month 1: 0.57
Churn rate at month 2: 0.21
Churn rate at month 3: -1.17
Churn rate at month 4: 0.05
Churn rate at month 5: -0.04
Churn rate at month 6: -0.04
Churn rate at month 7: 0.14
Churn rate at month 8: -0.02
Churn rate at month 9: -0.88
Churn rate at month 10: 0.31
Churn rate at month 11: -0.59
Churn rate at month 12: 0.10


In [118]:
prev_cnt = 46
for index, row in orders2016_cnt.iterrows():
    cnt = (prev_cnt - row['customer_id'])/prev_cnt
    print(f'Churn rate at month {index + 1}: {cnt:.2f}')
    prev_cnt = row['customer_id']

Churn rate at month 1: 0.00
Churn rate at month 2: 0.09
Churn rate at month 3: -0.90
Churn rate at month 4: -0.04
Churn rate at month 5: -0.16
Churn rate at month 6: 0.06
Churn rate at month 7: 0.01
Churn rate at month 8: 0.03
Churn rate at month 9: -1.05
Churn rate at month 10: 0.46
Churn rate at month 11: -0.74
Churn rate at month 12: 0.05


Теперь можно посмотреть на churn_rate в 2016 и в 2017 году по месяцам. Как видно из данных, большое кол-во клиентов ушло в период с 2016-12 по 2017-01 (было 176 клиентов в конце 2016, стало 69 в начале 17 года). Однако по churn rate видно, что последующие месяцы 2017 года в целом были неплохие с точки зрения оттока клиентов (в 6 месяцах мы видим отрицательное значение churn rate, значит, кол-во клиентов даже увеличивалось, причем в 11 месяцу -> 0.59 по сравнению с 10 месяцем; на 3 месяце -> 1.17 по сравнению со 2-м). Я бы выдвинула следующую гипотезу: проблема с доставкой всё-таки повлияла на отток пользователей в начале года, однако на последующие месяцы она влияния, вероятно, не оказывала (или улучшился какой-то другой фактор/ люди свыклись с такой доставкой). Кроме того, максимальное кол-во клиентов в 2017 году больше максимального кол-ва клиента в 2016. Видно, что 9 и 11 месяцы хорошие с точки зрения притока клиентов, и значения в 2017 году даже лучше значений 2016-го года. Для более глубокого анализа можно ещё изучить retention rate 2016 и 2017 года и сравнить значения